### ISSUE TITLE
eth_signTypedData on Trezor account fails without callback

### ISSUE THREAD

**Issue by rudolfix** [MetaMask/metamask-extension#5052]
When sending `eth_signTypedData` to web3 provider injected by metamask when Trezor account is active, callback (after clicking SIGN in ui) is not called and the console logs `Not supported on this device`.

to reproduce
1. add and select Trezor account
2. send properly formatted eth_signTypedData to the web3.currentProvider
3. Click SIGN in the signature ui

expected behavior
1. callback is being called in case of error
2. you are able to sign Ethereum message with Trezor

browse:
- Chrome 68.0.3440.84/Windows
- Metamask 4.9.2

This bug prevents users from logging in to services that use Ethereum wallets to login users. error message looks like Trezor does not support signing Messages, is this the case?

**Comment by bdresser** [comment]
@rudolfix Trezor doesn't currently support `eth_signTypedData`: https://github.com/trezor/trezor-core/issues/122

Since EIP 712 has been finalized, hopefully they'll begin to support it soon.

**Comment by rudolfix** [comment]
@bdresser thanks! what about eth_sign? I think that it is deprecated in Metamask so there is no way to use it, correct? do you known if any signature scheme works for Metamask + Trezor combination?

On the other note, calling a callback in case of error would be really helpful.

**Comment by bdresser** [comment]
Try `personal_sign`!

**Comment by rudolfix** [comment]
is there a way to detect we have trezor attached? I do not want to drop eth_signTypedData for regular accounts. anyway, thanks! we'll surely try `personal_sign` just to see if it works

**Comment by bdresser** [comment]
no, we don't expose whether the user is signing with a Trezor or a standard MetaMask account.

Since EIP 712 is merged, Trezor may hopefully support signTypedData shortly. Try commenting on the issue referenced above to let them know you're looking forward to this feature!

**Comment by danfinlay** [comment]
I have a new proposal for how we may be able to address this issue in the future:

Rather than allowing sites to detect the account type a user has, it could specify _before connecting_ what features it requires during the session:

https://github.com/MetaMask/metamask-extension/issues/6994

**Comment by xlab** [comment]
Support for EIP-712 in Trezor is tracked there:
https://github.com/trezor/trezor-firmware/pull/1568

**Comment by kargakis** [comment]
It seems like Trezor T support is merged: https://github.com/trezor/trezor-firmware/pull/1835

Are there any plans to add support in Trezor One?

**Comment by xlab** [comment]
T support can be implemented in the same way as in Ledger Nano S/X basically, a hash confirmation. I was waiting until Model One implementation will reach its destination. T impl is not a big deal.

**Comment by saluthejanitor** [comment]
> No immediate plans to add support for Trezor One: trezor/trezor-firmware#131 (comment)

@prusnak please consider adding EIP-712 support in Trezor One, buying a Trezor T just to use `eth_signTypedData` is not an answer.
